tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,310 @@
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let mut values = HashMap::new();
+        let mut visitor = RegistryVisitor(&mut values);
+        attrs.record(&mut visitor);
+        let parent = attrs.parent().map(|id| id.clone());
+        let s = Data {
+            metadata: attrs.metadata(),
+            parent,
+            children: vec![],
+            ref_count: AtomicUsize::new(1),
+            values: Mutex::new(values),
+        };
+        let id = self.insert(s).expect("Unable to allocate another span");
+        idx_to_id(id)
+    }
+
+    #[inline]
+    fn record(&self, span: &span::Id, values: &span::Record<'_>) {
+        dbg!(span);
+        dbg!(values);
+    }
+
+    fn record_follows_from(&self, span: &span::Id, follows: &span::Id) {
+        let current = self
+            .get(span)
+            .expect("Current span is missing; this is a bug");
+
+        let current = self
+            .span(span)
+            .expect("Current span is missing; this is a bug");
+    }
+
+    fn enter(&self, id: &span::Id) {
+        CURRENT_SPANS.with(|spans| {
+            spans.borrow_mut().push(id.clone());
+            if let Some(data) = self.get(id) {
+                data.ref_count.fetch_add(1, Ordering::SeqCst);
+            } else {
+                panic!("Span with ID {:?} was not found. This is a bug.", id);
+            }
+        })
+    }
+
+    fn event(&self, event: &Event<'_>) {
+        let id = match event.parent() {
+            Some(id) => Some(id.clone()),
+            None => {
+                if event.is_contextual() {
+                    CURRENT_SPANS.with(|spans| spans.borrow().current().map(|id| id.clone()))
+                } else {
+                    None
+                }
+            }
+        };
+        if let Some(id) = id {
+            let mut values = HashMap::new();
+            let mut visitor = RegistryVisitor(&mut values);
+            event.record(&mut visitor);
+            let _ = self.get(&id).expect("Missing parent span for event");
+        }
+    }
+
+    fn exit(&self, id: &span::Id) {
+        CURRENT_SPANS.with(|spans| {
+            spans.borrow_mut().pop(id);
+            if let Some(data) = self.get(id) {
+                data.ref_count.fetch_sub(1, Ordering::SeqCst);
+            } else {
+                panic!("Span with ID {:?} was not found", id);
+            }
+        })
+    }
+
+    fn clone_span(&self, id: &span::Id) -> span::Id {
+        if let Some(data) = self.get(&id) {
+            let refs = data.ref_count.fetch_add(1, Ordering::Release);
+            if refs == 0 {
+                panic!("tried to clone a span that already closed");
+            }
+        } else {
+            panic!("tried to clone {:?}, but no span exists with that ID", id)
+        }
+        id.clone()
+    }
+
+    #[inline]
+    fn try_close(&self, id: span::Id) -> bool {
+        self.drop_span(id)
```

Comment:
Why not just put all the logic in the separate `drop_span` method here? I don't understand why we have both?

Reply:
I moved this logic to the `Registry` which _should_ do the trick and reduce duplication.

Reply:
right, what I was getting at is that `Registry::drop_span` is only ever called by this method, why does there need to be a separate method? Why not just put the body of `drop_span` here?